### PR TITLE
[DROOLS-5381] Implemented code-generation for TreeModel

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @jiripetrlik @mariofusco 
See https://issues.redhat.com/browse/DROOLS-5381

1) `KiePMMLDroolsModel` returned by  `DroolsModelProvider` are code-generated
2) Concrete _DroolsModelProviders_ (e.g. `TreeModelImplementationProvider`) have to implement 
`Map<String, String> getKiePMMLDroolsModelSourcesMap(....)`
3) Most code for code-generation is shared for different `KiePMMLDroolsModel`s, so it has been put inside `KiePMMLDroolsModelFactoryUtils.getKiePMMLModelCompilationUnit(...)`
4) differences between constructors (e.g. super invocation) are managed locally by every implementation

### CODE REVIEW

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                                                                 final String modelClassName) {
+        logger.trace("getKiePMMLModelCompilationUnit {} {} {}", dataDictionary, model, packageName);
+        String className = getSanitizedClassName(model.getModelName());
+        String targetField = getTargetFieldName(dataDictionary, model).orElse(null);
+        List<KiePMMLOutputField> outputFields = getOutputFields(model);
+        CompilationUnit templateCU = getFromFileName(javaTemplate);
+        CompilationUnit cloneCU = templateCU.clone();
+        cloneCU.setPackageDeclaration(packageName);
+        ClassOrInterfaceDeclaration modelTemplate = cloneCU.getClassByName(modelClassName)
+                .orElseThrow(() -> new KiePMMLException(MAIN_CLASS_NOT_FOUND));
```
Comment (danielezonca) [inline kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java]: Can you please add `modelClassName` to error message?

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+                                   MethodCallExpr rankScopeExpr = (MethodCallExpr) valueScopeExpr.getScope().get();
+                                   expectedArguments = NodeList.nodeList(new IntegerLiteralExpr(outputField.getRank()));
+                                   toReturn &= commonEvaluateMethodCallExpr(rankScopeExpr, "withRank", expectedArguments, MethodCallExpr.class);
+                                   MethodCallExpr builderScopeExpr = (MethodCallExpr) rankScopeExpr.getScope().get();
+                                   expectedArguments = NodeList.nodeList(new StringLiteralExpr(outputField.getName()), new NameExpr("Collections.emptyList()"));
+                                   toReturn &= commonEvaluateMethodCallExpr(builderScopeExpr, "builder", expectedArguments, NameExpr.class);
+                                   toReturn &= builderScopeExpr.getName().equals(new SimpleName("builder"));
+                                   toReturn &= builderScopeExpr.getScope().get().equals(new NameExpr("KiePMMLOutputField"));
+                                   return toReturn;
```
Comment (danielezonca) [inline kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java]: Can you please move this logic to a method to improve code readability?

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                                   boolean toReturn = arguments.size() == 2;
+                                   Expression firstArgument = arguments.get(0);
+                                   Expression secondArgument = arguments.get(1);
+                                   toReturn &= firstArgument.isStringLiteralExpr() && ((StringLiteralExpr) firstArgument).getValue().equals(entry.getKey());
+                                   toReturn &= secondArgument.isObjectCreationExpr() &&
+                                           ((ObjectCreationExpr) secondArgument).getArgument(0).isStringLiteralExpr() &&
+                                           ((StringLiteralExpr) ((ObjectCreationExpr) secondArgument).getArgument(0)).getValue().equals(entry.getValue().getOriginalType()) &&
+                                           ((ObjectCreationExpr) secondArgument).getArgument(1).isStringLiteralExpr() &&
+                                           ((StringLiteralExpr) ((ObjectCreationExpr) secondArgument).getArgument(1)).getValue().equals(entry.getValue().getGeneratedType());
+                                   return toReturn;
```
Comment (danielezonca) [inline kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java]: Can you please move this logic to a method to improve code readability?

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        return statementStream
+                .filter(Statement::isExpressionStmt)
+                .map(expressionStmt -> ((ExpressionStmt) expressionStmt).getExpression())
+                .filter(expression -> expression instanceof MethodCallExpr)
+                .map(expression -> (MethodCallExpr) expression)
+                .filter(methodCallExpr ->
+                                methodCallExpr.getScope().isPresent() &&
+                                        methodCallExpr.getScope().get().isNameExpr() &&
+                                        ((NameExpr) methodCallExpr.getScope().get()).getName().asString().equals(scope) &&
+                                        methodCallExpr.getName().asString().equals(method))
```
Comment (danielezonca) [inline kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java]: Can you please move this logic to a method to improve code readability?
(minor note, fix indent)

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-compiler/src/main/java/org/kie/pmml/models/drools/scorecard/compiler/factories/KiePMMLScorecardModelFactory.java [hunk 5]
```diff
@@ -79,22 +65,14 @@ public static KiePMMLScorecardModel getKiePMMLScorecardModel(DataDictionary data
...
-        List<KiePMMLOutputField> outputFields = getOutputFields(model);
-        CompilationUnit templateCU = getFromFileName(KIE_PMML_SCORECARD_MODEL_TEMPLATE_JAVA);
-        CompilationUnit cloneCU = templateCU.clone();
-        cloneCU.setPackageDeclaration(packageName);
-        ClassOrInterfaceDeclaration modelTemplate = cloneCU.getClassByName(KIE_PMML_SCORECARD_MODEL_TEMPLATE)
-                .orElseThrow(() -> new RuntimeException(MAIN_CLASS_NOT_FOUND));
-        modelTemplate.setName(className);
-        MINING_FUNCTION miningFunction = MINING_FUNCTION.byName(model.getMiningFunction().value());
+        ClassOrInterfaceDeclaration modelTemplate = cloneCU.getClassByName(className)
+                .orElseThrow(() -> new KiePMMLException(MAIN_CLASS_NOT_FOUND));
```
Comment (danielezonca) [inline kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-compiler/src/main/java/org/kie/pmml/models/drools/scorecard/compiler/factories/KiePMMLScorecardModelFactory.java]: Can you please add `className` to error message?

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/factories/KiePMMLTreeModelFactory.java [hunk 6]
```diff
@@ -40,19 +48,38 @@
...
+
+    public static Map<String, String> getKiePMMLTreeModelSourcesMap(final DataDictionary dataDictionary,
+                                                                    final TreeModel model,
+                                                                    final Map<String, KiePMMLOriginalTypeGeneratedType> fieldTypeMap,
+                                                                    final String packageName) {
+        logger.trace("getKiePMMLTreeModelSourcesMap {} {} {}", dataDictionary, model, packageName);
+        CompilationUnit cloneCU = getKiePMMLModelCompilationUnit(dataDictionary, model, fieldTypeMap, packageName, KIE_PMML_TREE_MODEL_TEMPLATE_JAVA, KIE_PMML_TREE_MODEL_TEMPLATE);
+        String className = getSanitizedClassName(model.getModelName());
+        ClassOrInterfaceDeclaration modelTemplate = cloneCU.getClassByName(className)
+                .orElseThrow(() -> new KiePMMLException(MAIN_CLASS_NOT_FOUND));
```
Comment (danielezonca) [inline kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/factories/KiePMMLTreeModelFactory.java]: Can you please add `className` to error message?

Comment (gitgabrio) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/factories/KiePMMLTreeModelFactory.java]: @danielezonca 
done

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-compiler/src/main/java/org/kie/pmml/models/drools/scorecard/compiler/factories/KiePMMLScorecardModelFactory.java (hunk 5 above)
Comment (gitgabrio) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-compiler/src/main/java/org/kie/pmml/models/drools/scorecard/compiler/factories/KiePMMLScorecardModelFactory.java]: @danielezonca 
done

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java (hunk 4 above)
Comment (gitgabrio) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java]: @danielezonca 
done

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java (hunk 3 above)
Comment (gitgabrio) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java]: @danielezonca 
done

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java (hunk 2 above)
Comment (gitgabrio) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtilsTest.java]: @danielezonca 
done

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java (hunk 1 above)
Comment (gitgabrio) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java]: @danielezonca 
done

**Review by danielezonca — APPROVED** [review]

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+import static org.kie.pmml.compiler.commons.utils.ModelUtils.getTargetFieldName;
+
+/**
+ * Utility class to provide common methods for KiePMMLDroolsModel-specific <b>factories</b>
+ */
+public class KiePMMLDroolsModelFactoryUtils {
+
+    private static final Logger logger = LoggerFactory.getLogger(KiePMMLDroolsModelFactoryUtils.class.getName());
+
+    public KiePMMLDroolsModelFactoryUtils() {
```
Comment (jiripetrlik) [inline kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java]: Does this really avoid to create instances? It seems like public constructor.

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/executor/TreeModelImplementationProvider.java [hunk 8]
```diff
@@ -35,29 +34,27 @@
  */
 public class TreeModelImplementationProvider extends DroolsModelProvider<TreeModel, KiePMMLTreeModel> {
 
-    private static final Logger logger = LoggerFactory.getLogger(TreeModelImplementationProvider.class.getName());
```
Comment (jiripetrlik) [inline kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/executor/TreeModelImplementationProvider.java:38]: What is the reason for removing logging from this class? I thing logging on trace level is fine.

**Review by jiripetrlik — COMMENTED** [review]
I have two minor comments. Otherwise looks good.

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java (hunk 7 above)
Comment (gitgabrio) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/main/java/org/kie/pmml/models/drools/utils/KiePMMLDroolsModelFactoryUtils.java]: @jiripetrlik 
My bad! Thanks for spotting that

File: kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/executor/TreeModelImplementationProvider.java (hunk 8 above)
Comment (gitgabrio) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/executor/TreeModelImplementationProvider.java:38]: @jiripetrlik 
I'm trying to clean up things a little bit - all those "trace" were actually for "development" purpose, but I can not see a reason to have them when the code is merged.
I would prefer to eventually add them if needed, but then at "debug" level

Comment (jiripetrlik) [reply kie-pmml-new/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/executor/TreeModelImplementationProvider.java:38]: @gitgabrio 
I think in general it is better to have as many logs as possible if these logs are in the proper level. User can specify which level he needs and avoid of unnecessary logs in production. But I think it does not block this PR and we can merge.

**Comment by jiripetrlik** [comment]
@gitgabrio 
Thank you for the changes! Looks good to me now.

**Review by jiripetrlik — APPROVED** [review]

**Comment by gitgabrio** [comment]
@danielezonca Hi! Could you please merge this, squashing commits? Thanks!

**Comment by gitgabrio** [comment]
@danielezonca Hi! Could you please merge this, squashing commits? Thanks!

**Pull request merged**